File: Java/LinkedList/03_Singly_Circular_LinkedList_Java/SinglyCL_014.java
/*
================================================================================
Program Name    : DeleteFirst() and DeleteLast() in Singly Circular Linked List
Program Number  : 014
Description     : This program demonstrates the DeleteFirst() and DeleteLast()
                  operations in a Singly Circular Linked List using Java.
                  The program inserts nodes at both the beginning and end,
                  displays the list, deletes the first node, deletes the last
                  node, and updates the circular links along with the node count.
Date            : July 2026
Language        : Java
Author          : Ishwari Vijaykumar Surve
================================================================================
Concepts Used:
1. Object-Oriented Programming (OOP)
2. Class and Object
3. Constructor
4. Self-Referential Class
5. Singly Circular Linked List
6. InsertFirst()
7. InsertLast()
8. DeleteFirst()
9. DeleteLast()
10. Display Traversal using do-while Loop
11. Node Counting

Key Learning Points:
- DeleteFirst() removes the first node and updates first.
- DeleteLast() removes the last node by traversing to the second-last node.
- last.next always points to first to maintain circularity.
- Special cases of empty list and single node are handled.
- iCount is decremented after every successful deletion.

================================================================================
*/
// Program 14 : DeleteFirst() and DeleteLast() in Singly Circular Linked List
class node
{
    public int data;
    public node next;

    node(int no)
    {
        this.data = no;
        this.next = null;
    }
}

class SinglyCL
{
    public node first;
    public node last;
    public int iCount;

    public SinglyCL()
    {
        this.first = null;
        this.last = null;
        this.iCount = 0;
    }

    ///////////////////////////////////////////////////////////////

    public void Display()
    {
        node temp = first;

        if(first == null && last == null)
        {
            return;
        }

        do
        {
            System.out.print("| " + temp.data + " | -> ");
            temp = temp.next;
        }while(temp != last.next);

        System.out.println();
    }

    ///////////////////////////////////////////////////////////////

    public int Count()
    {
        return iCount;
    }

    ///////////////////////////////////////////////////////////////

    public void InsertFirst(int no)
    {
        node newn = new node(no);

        if(first == null && last == null)
        {
            first = newn;
            last = newn;
        }
        else
        {
            newn.next = first;
            first = newn;
        }

        last.next = first;

        iCount++;
    }

    ///////////////////////////////////////////////////////////////

    public void InsertLast(int no)
    {
        node newn = new node(no);

        if(first == null && last == null)
        {
            first = newn;
            last = newn;
        }
        else
        {
            last.next = newn;
            last = newn;
        }

        last.next = first;

        iCount++;
    }

    ///////////////////////////////////////////////////////////////

    public void InsertAtPos(int no, int iPos)
    {

    }

    ///////////////////////////////////////////////////////////////

    public void DeleteFirst()
    {
        if(first == null && last == null)
        {
            return;
        }
        else if(first == last)
        {
            first = null;
            last = null;
        }
        else
        {
            first = first.next;
            last.next = first;
        }

        iCount--;
    }

    ///////////////////////////////////////////////////////////////

    public void DeleteLast()
    {
        if(first == null && last == null)
        {
            return;
        }
        else if(first == last)
        {
            first = null;
            last = null;
        }
        else
        {
            node temp = first;

            while(temp.next != last)
            {
                temp = temp.next;
            }

            last = temp;
            last.next = first;
        }

        iCount--;
    }

    ///////////////////////////////////////////////////////////////

    public void DeleteAtPos(int iPos)
    {

    }
}

///////////////////////////////////////////////////////////////////////////

class SinglyCL14
{
    public static void main(String A[])
    {
        SinglyCL sobj = new SinglyCL();
        int iRet = 0;

        sobj.InsertFirst(51);
        sobj.InsertFirst(21);
        sobj.InsertFirst(11);

        sobj.InsertLast(101);
        sobj.InsertLast(111);
        sobj.InsertLast(121);

        sobj.Display();

        iRet = sobj.Count();
        System.out.println("Number of nodes are : " + iRet);

        sobj.DeleteFirst();

        sobj.Display();

        iRet = sobj.Count();
        System.out.println("Number of nodes are : " + iRet);

        sobj.DeleteLast();

        sobj.Display();

        iRet = sobj.Count();
        System.out.println("Number of nodes are : " + iRet);
    }
}
/*
================================================================================
Expected Output:
================================================================================

| 11 | -> | 21 | -> | 51 | -> | 101 | -> | 111 | -> | 121 | ->
Number of nodes are : 6

| 21 | -> | 51 | -> | 101 | -> | 111 | -> | 121 | ->
Number of nodes are : 5

| 21 | -> | 51 | -> | 101 | -> | 111 | ->
Number of nodes are : 4

================================================================================
Memory Representation:
================================================================================

Initial Circular Linked List

                 first
                   │
                   ▼
        +----+      +----+      +----+      +-----+      +-----+      +-----+
        | 11 | ───► | 21 | ───► | 51 | ───► | 101 | ───► | 111 | ───► | 121 |
        +----+      +----+      +----+      +-----+      +-----+      +-----+
           ▲                                                             │
           └────────────────────────────── last.next ─────────────────────┘
                                                               ▲
                                                               │
                                                              last

After DeleteFirst()

                 first
                   │
                   ▼
        +----+      +----+      +-----+      +-----+      +-----+
        | 21 | ───► | 51 | ───► | 101 | ───► | 111 | ───► | 121 |
        +----+      +----+      +-----+      +-----+      +-----+
           ▲                                                  │
           └──────────────────── last.next ───────────────────┘
                                                     ▲
                                                     │
                                                    last

After DeleteLast()

                 first
                   │
                   ▼
        +----+      +----+      +-----+      +-----+
        | 21 | ───► | 51 | ───► | 101 | ───► | 111 |
        +----+      +----+      +-----+      +-----+
           ▲                                       │
           └────────────── last.next ──────────────┘
                                           ▲
                                           │
                                          last

================================================================================
Explanation:
================================================================================

The constructor initializes first and last to null and sets iCount to 0.

InsertFirst() inserts a new node at the beginning of the circular linked
list and updates last.next to maintain the circular connection.

InsertLast() inserts a new node at the end of the list, updates last to
the newly inserted node, and reconnects last.next to first.

Display() traverses the circular linked list using a do-while loop and
stops when the traversal again reaches the first node.

DeleteFirst() handles three cases:

• If the list is empty, no deletion is performed.
• If only one node exists, both first and last become null.
• Otherwise, first moves to the next node and last.next is updated.

DeleteLast() also handles three cases:

• If the list is empty, nothing is deleted.
• If only one node exists, both first and last become null.
• Otherwise, traversal reaches the second-last node, last is updated,
  and last.next is linked back to first.

The node counter (iCount) is decremented after every successful deletion.

================================================================================
Program Flow:
================================================================================

Create SinglyCL Object
          │
          ▼
Constructor Invoked
          │
          ▼
InsertFirst()
          │
          ▼
InsertLast()
          │
          ▼
Display Linked List
          │
          ▼
Count Nodes
          │
          ▼
DeleteFirst()
          │
          ▼
Display Updated List
          │
          ▼
DeleteLast()
          │
          ▼
Display Updated List
          │
          ▼
Program Terminates

================================================================================
Operations Performed:
================================================================================

1. Created a Singly Circular Linked List object.
2. Inserted nodes using InsertFirst().
3. Inserted nodes using InsertLast().
4. Displayed the circular linked list.
5. Counted the total number of nodes.
6. Deleted the first node.
7. Displayed the updated linked list.
8. Deleted the last node.
9. Displayed the final linked list.
10. Updated iCount after each deletion.

================================================================================
Time Complexity:
================================================================================

InsertFirst()   : O(1)
InsertLast()    : O(1)
DeleteFirst()   : O(1)
DeleteLast()    : O(n)
Display()       : O(n)
Count()         : O(1)

Space Complexity: O(1)

================================================================================
Conclusion:
================================================================================
Program 014 demonstrates the DeleteFirst() and DeleteLast() operations in
a Singly Circular Linked List. The program correctly maintains the circular
link by ensuring last.next always points to first. It also handles empty,
single-node, and multiple-node cases while keeping the node count accurate,
making the implementation reliable and efficient.

================================================================================
*/

File: C++/LinkedList/03_Singly_Circular_Linear_LinkedList_CPP/SinglyCL_014.cpp
/*
===============================================================================
Program Name    : Singly Circular Linked List - DeleteLast Operation
Program Number  : 014
Description     : This program implements the DeleteLast() function for a
                  Singly Circular Linked List. It removes the last node,
                  updates the last pointer, reconnects the circular link,
                  and maintains the node count.

Date            : July 2026
Language        : C++
Author          : Ishwari Vijaykumar Surve
===============================================================================
Concepts Used:
1. Singly Circular Linked List
2. DeleteLast()
3. Dynamic Memory Deallocation
4. Pointer Traversal
5. Circular Link Maintenance
6. Class and Objects

Key Learning Points:
- Delete the last node of a circular linked list.
- Traverse to the second last node.
- Update the last pointer correctly.
- Maintain the circular link.
- Handle empty and single-node lists.

===============================================================================
*/
#include<iostream>
using namespace std;

#pragma pack(1)

struct node
{
    int data;
    struct node *next;
};

typedef struct node NODE;
typedef struct node *PNODE;

class SinglyCL
{
private:
    PNODE first;
    PNODE last;
    int iCount;

public:
    SinglyCL();

    void Display();
    int Count();

    void InsertFirst(int);
    void InsertLast(int);

    void DeleteFirst();
    void DeleteLast();

    void InsertAtPos(int,int){}
    void DeleteAtPos(int){}
};

SinglyCL::SinglyCL()
{
    first = NULL;
    last = NULL;
    iCount = 0;
}

void SinglyCL::Display()
{
    if(first == NULL && last == NULL)
    {
        return;
    }

    PNODE temp = first;

    do
    {
        cout<<"| "<<temp->data<<" | -> ";
        temp = temp->next;
    }while(temp != last->next);

    cout<<"\n";
}

int SinglyCL::Count()
{
    return iCount;
}

void SinglyCL::InsertFirst(int iNo)
{
    PNODE newn = new NODE;

    newn->data = iNo;
    newn->next = NULL;

    if(first == NULL && last == NULL)
    {
        first = last = newn;
    }
    else
    {
        newn->next = first;
        first = newn;
    }

    last->next = first;

    iCount++;
}

void SinglyCL::InsertLast(int iNo)
{
    PNODE newn = new NODE;

    newn->data = iNo;
    newn->next = NULL;

    if(first == NULL && last == NULL)
    {
        first = last = newn;
    }
    else
    {
        last->next = newn;
        last = newn;
    }

    last->next = first;

    iCount++;
}

void SinglyCL::DeleteFirst()
{
    if(first == NULL && last == NULL)
    {
        return;
    }
    else if(first == last)
    {
        delete first;
        first = NULL;
        last = NULL;
    }
    else
    {
        first = first->next;
        delete last->next;
        last->next = first;
    }

    iCount--;
}

void SinglyCL::DeleteLast()
{
    if(first == NULL && last == NULL)
    {
        return;
    }
    else if(first == last)
    {
        delete first;
        first = NULL;
        last = NULL;
    }
    else
    {
        PNODE temp = first;

        while(temp->next != last)
        {
            temp = temp->next;
        }

        delete last;
        last = temp;
        last->next = first;
    }

    iCount--;
}

int main()
{
    SinglyCL sobj;
    int iRet = 0;

    sobj.InsertFirst(51);
    sobj.InsertFirst(21);
    sobj.InsertFirst(11);

    sobj.InsertLast(101);
    sobj.InsertLast(111);
    sobj.InsertLast(121);

    sobj.Display();
    iRet = sobj.Count();
    cout<<"Number of nodes are : "<<iRet<<"\n";

    sobj.DeleteFirst();

    sobj.Display();
    iRet = sobj.Count();
    cout<<"Number of nodes are : "<<iRet<<"\n";

    sobj.DeleteLast();

    sobj.Display();
    iRet = sobj.Count();
    cout<<"Number of nodes are : "<<iRet<<"\n";

    return 0;
}
/*
===============================================================================
Expected Output:
===============================================================================

| 11 | -> | 21 | -> | 51 | -> | 101 | -> | 111 | -> | 121 |

Number of nodes are : 6

| 21 | -> | 51 | -> | 101 | -> | 111 | -> | 121 |

Number of nodes are : 5

| 21 | -> | 51 | -> | 101 | -> | 111 |

Number of nodes are : 4

-Before DeleteLast()
===============================================================================
Memory Representation:
===============================================================================

first                                              last
  │                                                  │
  ▼                                                  ▼

+------+ --> +------+ --> +------+ --> +------+ --> +------+ --> +------+
| 21   |     | 51   |     |101   |     |111   |     |121   |     |131   |
+------+     +------+     +------+     +------+     +------+     +------+
    ^                                                                     |
    |_____________________________________________________________________|

DeleteLast()

-After DeleteLast()
first                                   last
  │                                       │
  ▼                                       ▼

+------+ --> +------+ --> +------+ --> +------+
| 21   |     | 51   |     |101   |     |111   |
+------+     +------+     +------+     +------+
    ^                                                   |
    |___________________________________________________|

Deleted Node : 121

===============================================================================
Algorithm:
===============================================================================

DeleteLast()

1. Check whether the list is empty.
2. If only one node is present:
      Delete the node.
      Set first and last to NULL.
3. Otherwise:
      Traverse to the second last node.
      Delete the last node.
      Update last = second last node.
      Update last->next = first.
4. Decrement iCount.

===============================================================================
Program Flow:
===============================================================================

Create List
      │
      ▼
Insert Nodes
      │
      ▼
Display List
      │
      ▼
DeleteFirst()
      │
      ▼
Display List
      │
      ▼
DeleteLast()
      │
      ▼
Traverse to Second Last Node
      │
      ▼
Delete Last Node
      │
      ▼
Update last Pointer
      │
      ▼
Update last->next = first
      │
      ▼
Display Updated List

===============================================================================
Time Complexity:
===============================================================================

DeleteFirst() : O(1)

DeleteLast()  : O(n)

Space Complexity : O(1)

===============================================================================
Conclusion:
===============================================================================

Program 014 extends the Singly Circular Linked List by implementing the
DeleteLast() operation. The list is traversed to locate the second last
node, after which the last node is removed and the circular link is
restored. This program demonstrates proper pointer manipulation while
maintaining the integrity of the circular linked list.

===============================================================================
*/
